src/dcd/server/autocomplete/calltip_utils.d

```diff
@@ -0,0 +1,92 @@
+module dcd.server.autocomplete.calltip_utils;
+
+import std.string;
+import std.regex;
+import std.range : empty;
+import std.experimental.logger;
+import std.algorithm : canFind;
+
+/** 
+ * 
+ * Params:
+ *   callTip = the symbols calltip
+ * Returns: the first argument type of the calltip 
+ */
+string getFirstArgumentOfFunction(string callTip)
+{
+    // First match all inside the parentheses
+    auto insideParenthesesRegex = regex(`\((.*\))`);
+    auto match = matchFirst(callTip, insideParenthesesRegex);
+    string insideParentheses = match.captures.back;
+
+    if (insideParentheses.empty)
+    {
+        return "";
+    }
+
+    // Then match the first word that isn't lazy return scope ... etc.
+    auto firstWordRegex = regex(`(?!lazy|return|scope|in|out|ref|const|immutable\b)\b\w+`);
+
+    auto matchFirstType = matchFirst(insideParentheses, firstWordRegex);
+    string firstArgument = matchFirstType.captures.back;
+    return firstArgument.empty ? "" : firstArgument;
+
+}
+
+string removeFirstArgumentOfFunction(string callTip)
+{
+    auto parentheseSplit = callTip.split('(');
+    // has only one argument
+    if (!callTip.canFind(','))
+    {
+        return parentheseSplit[0] ~ "()";
+    }
+    auto commaSplit = parentheseSplit[1].split(',');
+    string newCallTip = callTip.replace((commaSplit[0] ~ ", "), "");
+    return newCallTip;
+
+}
+
+unittest
+{
+    auto result = getFirstArgumentOfFunction("void fooFunction(ref const(Foo) bar)");
```

Comment:
what about templates?

e.g.
```d
void fooFunction(T)(ref const(T) bar)
```

would match the (T) because it's using matchFirst.

Additionally these generic types don't work at all with UFCS and this pretty much needs to be entirely redone when we want to support templated arguments (which raises the mental threshold a lot)

IMO instead of parsing the callTip, you should store the first argument as symbol on the dsymbol side and best to also add a helper function to check if it's matching. This isn't something that should be done entirely on the DCD side.

Reply:
Honestly I didn't think about supporting template types, since I just focused on tackling implementing UFCS for concrete types.

Since it won't match we wont get it in our completion list, so there isn't any issues. 
We can say templates types isn't supported yet, and  probably isn't trivial to implement support for as you mentioned.

Indeed I like you solution, It would be nice if dsymbol just was a part of dcd.
For now we just rely on the calltip which will be replaced by a better solution hopefully in the future.